File: firmware/core2/src/domain/ranking.cpp
#include "domain/ranking.h"

#include <algorithm>
#include <cstring>

#include "domain/geo.h"

namespace ofd {

AircraftList rankNearest(const AircraftList& input, const CircleMonitoringArea& area, size_t maxResults) {
  AircraftList result;

  // Stage into a temporary buffer with distance/bearing filled in, then
  // sort, then copy the bounded head into the output -- avoids any
  // dynamic allocation.
  AircraftState staged[kMaxAircraftPerUpdate];
  size_t stagedCount = 0;

  for (size_t i = 0; i < input.count && stagedCount < kMaxAircraftPerUpdate; i++) {
    const AircraftState& a = input.items[i];
    if (!isWithinCircle(a.latitude, a.longitude, area.centerLat, area.centerLon, area.radiusKm)) {
      continue;
    }
    AircraftState enriched = a;
    enriched.distanceFromObserverKm =
        haversineDistanceKm(area.centerLat, area.centerLon, a.latitude, a.longitude);
    enriched.hasDistanceFromObserverKm = true;
    enriched.bearingFromObserverDeg =
        initialBearingDeg(area.centerLat, area.centerLon, a.latitude, a.longitude);
    enriched.hasBearingFromObserverDeg = true;
    staged[stagedCount++] = enriched;
  }

  std::sort(staged, staged + stagedCount, [](const AircraftState& a, const AircraftState& b) {
    return a.distanceFromObserverKm < b.distanceFromObserverKm;
  });

  const size_t bound = std::min(maxResults, stagedCount);
  for (size_t i = 0; i < bound; i++) {
    result.items[i] = staged[i];
  }
  result.count = bound;
  return result;
}

}  // namespace ofd

File: services/gateway/src/lib/ranking.ts
import type { AircraftState, CircleMonitoringArea } from "@openflightdisplay/shared-models";
import { MAX_AIRCRAFT_PER_UPDATE } from "@openflightdisplay/shared-models";
import { haversineDistanceKm, initialBearingDeg, isWithinCircle } from "./geo.js";

/** Aircraft older than this are flagged stale rather than presented as live. */
export const STALE_POSITION_THRESHOLD_MS = 60_000;

export function isStale(aircraft: Pick<AircraftState, "positionTimestamp">, now: Date): boolean {
  return now.getTime() - Date.parse(aircraft.positionTimestamp) > STALE_POSITION_THRESHOLD_MS;
}

/**
 * The only ranking mode implemented in Phase 1: nearest horizontal
 * distance to the monitoring area's center, restricted to aircraft
 * actually inside the circle. Other modes (slant range, closest
 * approach, highest/lowest/fastest, weighted relevance, ...) are
 * documented in docs/FEATURE_PARITY_MATRIX.md and deliberately not
 * implemented yet.
 */
export function rankNearest(
  aircraftList: readonly AircraftState[],
  area: CircleMonitoringArea,
  maxResults: number = MAX_AIRCRAFT_PER_UPDATE,
): AircraftState[] {
  const enriched = aircraftList
    .filter((a) => isWithinCircle(a.latitude, a.longitude, area.centerLat, area.centerLon, area.radiusKm))
    .map((a) => {
      const distanceFromObserverKm = haversineDistanceKm(
        area.centerLat,
        area.centerLon,
        a.latitude,
        a.longitude,
      );
      const bearingFromObserverDeg = initialBearingDeg(
        area.centerLat,
        area.centerLon,
        a.latitude,
        a.longitude,
      );
      return { ...a, distanceFromObserverKm, bearingFromObserverDeg };
    });

  enriched.sort((a, b) => a.distanceFromObserverKm! - b.distanceFromObserverKm!);
  return enriched.slice(0, maxResults);
}
